src/transaction/transaction.rs

```diff
@@ -28,9 +37,104 @@ use std::{collections::BTreeMap, ops::RangeBounds};
 /// ```
 #[derive(new)]
 pub struct Transaction {
-    pub snapshot: Snapshot,
+    pub timestamp: Timestamp,
     #[new(default)]
-    mutations: BTreeMap<Key, Mutation>,
+    buffer: Buffer,
+    rpc: Arc<PdRpcClient>,
+}
+
+#[derive(Default)]
+struct Buffer {
+    mutations: Mutex<BTreeMap<Key, Mutation>>,
+}
+
+impl Buffer {
+    async fn get_or_else<F, Fut>(&self, key: Key, f: F) -> Result<Option<Value>>
+    where
+        F: FnOnce(Key) -> Fut,
+        Fut: Future<Output = Result<Option<Value>>>,
+    {
+        match self.get_from_mutations(&key) {
+            MutationValue::Determined(value) => Ok(value),
+            MutationValue::Undetermined => {
+                let value = f(key.clone()).await?;
+                let mut mutations = self.mutations.lock().unwrap();
+                mutations.insert(key, Mutation::Cached(value.clone()));
+                Ok(value)
+            }
+        }
+    }
+
+    async fn batch_get_or_else<F, Fut>(
```

Comment:
I hope the order can be retained if we . You may refer to my pr: https://github.com/tikv/client-rust/pull/100/files#diff-2dca5666975e7d44fe0f80bcd94c551aR412

Reply:
I don't think it is necessary - the result should be treated like a map, not an ordered list. If we had a `Map` type in Rust I would use it

Reply:
So why not a `HashMap`?

Reply:
By returning the iterator we give the user the choice about whether to use a BTreeMap or HashMap, we'd only collect the iterator into a HashMap anyway. Having said that, I'd be fine with this as a HashMap, tbh.